src/adapter/src/catalog/apply.rs

```diff
@@ -1009,25 +1123,68 @@ impl CatalogState {
     }
 }
 
-/// Inserts `key` and `value` into `map` if `diff` is an addition, otherwise remove them from `map`
-/// if `diff` is a retraction.
-fn apply<K, V>(map: &mut BTreeMap<K, V>, key: K, value: impl FnOnce() -> V, diff: StateDiff)
+/// Helper method to updated inverted lookup maps. The keys are generally names and the values are
+/// generally IDs.
+///
+/// Importantly, when retracting it's expected that the existing value will match `value` exactly.
+fn apply_inverted_lookup<K, V>(map: &mut BTreeMap<K, V>, key: &K, value: V, diff: StateDiff)
 where
-    K: Ord + Debug,
+    K: Ord + Clone + Debug,
     V: PartialEq + Debug,
 {
     match diff {
         StateDiff::Retraction => {
-            let prev = map.remove(&key);
-            // We can't assert the exact contents of the previous value, since we don't know
-            // what it should look like.
-            assert!(
-                prev.is_some(),
+            let prev = map.remove(key);
+            assert_eq!(
+                prev,
+                Some(value),
                 "retraction does not match existing value: {key:?}"
             );
         }
         StateDiff::Addition => {
-            let prev = map.insert(key, value());
+            let prev = map.insert(key.clone(), value);
+            assert_eq!(
+                prev, None,
+                "values must be explicitly retracted before inserting a new value: {key:?}"
+            );
+        }
+    }
+}
+
+/// Helper method to update catalog state, that may need to be updated from a previously retracted
+/// object.
+fn apply_with_update<K, V, D>(
+    map: &mut BTreeMap<K, V>,
+    durable: D,
+    key_fn: impl FnOnce(&D) -> K,
+    diff: StateDiff,
+    retractions: &mut BTreeMap<D::Key, V>,
+) where
+    K: Ord,
+    V: UpdateFrom<D> + PartialEq + Debug,
+    D: DurableType,
+    D::Key: Ord,
+{
+    match diff {
+        StateDiff::Retraction => {
+            let mem_key = key_fn(&durable);
+            let value = map
+                .remove(&mem_key)
+                .expect("retraction does not match existing value: {key:?}");
+            let durable_key = durable.into_key_value().0;
+            retractions.insert(durable_key, value);
+        }
+        StateDiff::Addition => {
+            let mem_key = key_fn(&durable);
+            let durable_key = durable.key();
+            let value = match retractions.remove(&durable_key) {
+                Some(mut value) => {
+                    value.update_from(durable);
```

Comment:
How does `update_from` not squash any value that were changed in the new object?

Reply:
I might be misunderstanding the question. `value` is the old object, `durable` is the new object. So we're taking all the fields from the new object and moving them into the old object. I changed the name of `value` to `retraction`, does that help at all?

Reply:
Ahh, yes. Thanks.